packages/analytics/amplify_analytics_pinpoint_dart/test/queued_item_store_test.dart

```diff
@@ -0,0 +1,244 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import 'package:amplify_analytics_pinpoint_dart/src/impl/analytics_client/event_client/queued_item_store/dart_queued_item_store.dart';
+import 'package:amplify_db_common_dart/amplify_db_common_dart.dart';
+
+import 'package:test/test.dart';
+
+void main() {
+  late DartQueuedItemStore db;
+
+  group('DartQueuedItemStore Tests', () {
+    setUpAll(() {
+      final qe = connect(name: 'test', path: '.');
+      db = DartQueuedItemStore(qe);
+    });
+
+    /// Reload db instance
+    Future<void> refreshDb() async {
+      await db.clear();
+    }
+
+    setUp(() async {
+      await refreshDb();
+    });
+
+    tearDown(() async {
+      await db.clear();
+    });
+
+    Future<void> validateStoredValues(
+      List<String> targetValues, {
+      int numToGet = 100,
+    }) async {
+      final readElements = await db.get(numToGet);
+
+      expect(readElements.length, targetValues.length);
+      for (var i = 0; i < readElements.length; i++) {
+        expect(
+          reason: 'Comparing element $i of readElements vs targetValues',
+          readElements.elementAt(i).value,
+          targetValues[i],
+        );
+      }
+    }
+
+    Future<void> writeValues(List<String> values) async {
+      for (final value in values) {
+        await db.add(value);
+      }
+    }
+
+    Future<void> deleteValues(List<String> valuesToDelete) async {
+      final itemsToDelete = (await db.get(100)).where((item) {
+        return valuesToDelete.contains(item.value);
+      });
+      await db.deleteItems(itemsToDelete);
+    }
+
+    test('Values stored', () async {
+      const values = ['0', '1', '2', '3', '4', '5'];
+      await writeValues(values);
+
+      await validateStoredValues(numToGet: 6, values);
+    });
+
+    test('Only requested values are returned', () async {
+      const values = ['0', '1', '2', '3', '4', '5'];
+      await writeValues(values);
+
+      await validateStoredValues(numToGet: 3, values.sublist(0, 3));
+    });
```

Comment:
These tests are still very dry. I have to look at writeValues and validateStoredValues to understand the tests. I'm not sure what APIs are being tested without digging into those functions.

The test names aren't helping me understand the test either. "Only requested values are returned" doesn't really explain the expected behavior. The expectation is that this behaves like a queue. That is, the first n items are returned.

Below is an example of how I might re-write the first two tests.

```dart
test('writes values to storage', () async {
  const values = ['0', '1', '2', '3', '4', '5'];
  for (final value in values) {
    await db.add(value);
  }

  final storedItems = await db.get(6);
  final storedValues = storedItems.map((e) => e.value);

  expect(storedValues, values);
});

test('returns first n items in the queue', () async {
  const values = ['0', '1', '2', '3', '4', '5'];
  for (final value in values) {
    await db.add(value);
  }

  final storedItems = await db.get(3);
  final storedValues = storedItems.map((e) => e.value);

  expect(storedValues, values.take(3));
});
```

Reply:
Based on the DAMP vs. DRY link, the purpose of code duplication is to "promote readability".  The tradeoff seemed to be described as conciseness vs readability.  

I don't believe that the remaining method abstractions I have reduce readability, instead I think they actively increase readability.  

Beyond the three method abstractions (writeValues, validateStoredValues, deleteValues) there aren't any other helper methods.  Furthermore each method does a basic task that can be inferred from its name. 

When reading code isn't it helpful to quickly know the intent of what's happening?  When reading the above code example, it seems harder to do. 

There are more redundant basic details that distract attention from understanding what's going on.  

While the intent is to validate the stored values, one has to figure that out by reading this code: 

```
final storedItems = await db.get(6);
  final storedValues = storedItems.map((e) => e.value);
  expect(storedValues, values);
```

Which could be: 
`      await validateStoredValues(numToGet: 6, values);
`

Once one knows the intent and sees that code pattern throughout every test, one still needs to parse x3 as much code to understand which distracts from understanding the test's purpose. 

What benefits do you see for expanding the code out in this way? 

I agree that the naming of the tests can be better though and have updated that.

Reply:
The tests as they are do not directly call any of the APIs they are testing. I don't know what APIs you are trying to test without looking into the underlying functions, which have far more code to parse through. Unit tests should generally directly call the APIs they are testing so it is clear what API is under test.